File: TwoPointer/3Sum.py
# https://leetcode.com/problems/3sum/
from typing import List
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        #-4 -1 -1 0 1 2
        nums.sort()
        result = []
        for left in range(len(nums)-2):
            if left > 0 and nums[left] == nums[left-1]:
                continue
            mid = left + 1
            right = len(nums) - 1

            while mid < right:
                summ = nums[left] + nums[mid] + nums[right]
                if summ < 0:
                    mid += 1
                elif summ > 0:
                    right -= 1
                else:
                    result.append([nums[left],nums[mid],nums[right]]) 
                    while mid < right and nums[mid] == nums[mid+1]:
                        mid+=1
                    while mid < right and nums[right] == nums[right-1]:
                        right -= 1

                    mid +=1
                    right -= 1
        return result


s = Solution()
print(s.threeSum([0,1,1]))  # Output: [[-1,-1,2],[-1,0,1]]

File: TwoPointer/3Sum.kt
// https://leetcode.com/problems/3sum/description/
fun main() {
    // Your code here
    val nums = intArrayOf(-1, 0, 1, 2, -1, -4)
    val result = threeSum(nums)
    println(result)
}

fun threeSum(nums: IntArray): List<List<Int>> {
        nums.sort()
        val result = mutableListOf<List<Int>>()
        for(i in 0 until nums.size - 2) {
            if(i > 0 && nums[i] == nums[i - 1]) continue
            var k = i + 1
            var j = nums.size - 1
            while(k < j) {
                val sum = nums[i] + nums[k] + nums[j]
                if (sum < 0) {
                    k++
                } else if (sum > 0) {
                    j--
                } else {
                    result.add(listOf(nums[i], nums[k], nums[j]))
                    while(k < j && nums[k] == nums[k + 1]) k++
                    while(k < j && nums[j] == nums[j - 1]) j--
                    k++
                    j--
                }
            }
        }
        return result
    }